File: 장세훈/Programmers/Level2/프렌즈4블록.java
class 프렌즈4블록 {

    /*
        https://school.programmers.co.kr/learn/courses/30/lessons/17679
     */

    public int solution(int m, int n, String[] board) {
        char[][] map = new char[m][n];
        int answer = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                map[i][j] = board[i].charAt(j);
            }
        }

        int count;
        do {
            count = brokePosition(map);
            answer += count;
        } while (count != 0);

        return answer;
    }


    int brokePosition(char[][] map) {
        Deque<int[]> queue = new ArrayDeque<>();
        int count = 0;

        for (int i = 0; i < map.length - 1; i++) {
            for (int j = 0; j < map[i].length - 1; j++) {
                if (map[i][j] != '0' && map[i][j] == map[i + 1][j] && map[i][j] == map[i][j + 1] && map[i][j] == map[i + 1][j + 1]) {
                    queue.add(new int[]{i, j});
                    queue.add(new int[]{i, j + 1});
                    queue.add(new int[]{i + 1, j});
                    queue.add(new int[]{i + 1, j + 1});
                }
            }
        }

        while (!queue.isEmpty()) {
            int[] poll = queue.poll();
            if(map[poll[0]][poll[1]] != '0') {
                map[poll[0]][poll[1]] = '0';
                count++;
            }
        }

        for (int i = 0; i < map[0].length; i++) {
            for (int j = map.length - 1; j > 0; j--) {
                if (map[j][i] == '0') {
                    int k = j;
                    while (k > -1 && map[k][i] == '0') {
                        k--;
                    }

                    if (k > -1) {
                        map[j][i] = map[k][i];
                        map[k][i] = '0';
                    }
                }

            }
        }

        return count;
    }

}

File: 정수빈/Programmers/Level2/프렌즈4블록.py
# https://school.programmers.co.kr/learn/courses/30/lessons/17679?language=python3

def solution(m, n, board):
    answer = 0
    directions = [(0,0), (1,0), (0,1), (1,1)]
    board = [list(x) for x in board]
    def drop():
        for c in range(n):
            for r in range(m-1, -1, -1):
                if board[r][c] == None:
                    i = r-1
                    while i >= 0:
                        if board[i][c] != None:
                            board[r][c], board[i][c] = board[i][c], board[r][c]
                            break
                        i-=1
    
    def disappear():
        diss = set()
        for r in range(m):
            for c in range(n):
                character = board[r][c]
                four = []
                for dr, dc in directions:
                    nr, nc = r+dr, c+dc
                    if 0<=nr<m and 0<=nc<n and board[nr][nc] == character:
                        four.append((nr, nc))
                    else:
                        break
                if len(four) == 4:
                    for f in four:
                        diss.add(tuple(f))
        # print(diss)
        for r, c in diss:
            board[r][c] = None

    def count_disappear():
        count = 0
        for r in range(m):
            for c in range(n):
                if board[r][c] == None:
                    count += 1
        return count
    

    prev_count = None
    count = 0
    while prev_count != count:
        disappear()
        drop()
        prev_count = count
        count = count_disappear()
    return count